nengo_spinnaker/netlist/utils.py

```diff
@@ -0,0 +1,169 @@
+import collections
+import rig.netlist
+from six import iteritems, itervalues
+
+from nengo_spinnaker.utils.keyspaces import is_nengo_keyspace
+
+
+def get_nets_for_placement(nets):
+    """Convert a list of N:M nets into a list of Rig nets suitable for
+    performing placement.
+
+    Parameters
+    ----------
+    nets : [:py:class:`~nengo_spinnaker.netlist.Net`, ...]
+        N:M nets to convert
+
+    Yields
+    ------
+    :py:class:`~rig.netlist.Net`
+        1:M net suitable for use with Rig for placement purposes.
+    """
+    # For each source in each net create a new Rig net
+    for net in nets:
+        for source in net.sources:
+            yield rig.netlist.Net(source, net.sinks, net.weight)
+
+
+def get_nets_for_routing(nets, placements):
+    """Convert a list of N:M nets into a list of Rig nets suitable for
+    performing routing.
+
+    Parameters
+    ----------
+    nets : [:py:class:`~nengo_spinnaker.netlist.Net`, ...]
+    placements : {vertex: (x, y), ...}
+
+    Returns
+    -------
+    [:py:class:`~rig.netlist.Net`, ...]
+        1:M net suitable for use with Rig for routing purposes.
+    {vertex: (x, y), ...}
+        An extended copy of the placements dictionary which must be used when
+        performing routing with the returned nets.
+    {:py:class:`~nengo_spinnaker.netlist.Net`:
+            {(x, y): :py:class:`~rig.netlist.Net`, ...}, ...}
+        Map from original nets to co-ordinates and the derived nets which
+        originate from them.
+    """
+    routing_nets = list()  # Nets with which to perform routing
+    extended_placements = dict(placements)  # New placements will be added
+    derived_nets = collections.defaultdict(dict)  # {Net: {placement: rig.Net}}
+
+    # For each Net build a set of all the co-ordinates from which the net now
+    # originates.
+    for net in nets:
+        start_placements = set(placements[v] for v in net.sources)
+
+        # For each of these co-ordinates create a new Rig Net with a new source
+        # vertex placed at the given co-ordinate.
+        for placement in start_placements:
+            # Create a new source vertex and place it at the given placement
+            vertex = object()
+            extended_placements[vertex] = placement
```

Comment:
I realise that current implementations (and probably future ones too) of the router do not actually look at `vertices_resources` and just placements for a list of vertices but strictly speaking letting these structures go out of sync is a bug.

I'd suggest one of:

1. A big scary comment and continue like you are
2. Update vertices_resources and allocations too
3. Use an arbitrarily selected existing vertex on the chip in question.

Up to you :).